Improve query API docs

Nilirad opened the pull request:
# Objective

- `Query` docs need a refresh and they look quite inconsistent.
- `WorldQuery` docs had some info about the derive macro that should not belong there.
- Closes #4742.

## Solution

Documentation has been changed for the following items:

- `Query` (item level and methods)
- `QueryCombinationIter`
- `WorldQuery` trait
- `WorldQuery` macro

### `Query`

- Introduction has been redone.
- “Renamed” type parameters to reflect their purpose.
- Examples for declaration idioms have been simplified and reordered.
- “Query result” terminology has been changed to “query item”, which is more in line with the underlying types and with general terminology.
- Query methods: structure has been made consistent, old notes about query being able to access certain methods based on mutability have been updated and centralized to the item level.
- A tables has been added to navigate the safe methods.
- A performance section has been added, notably with a table with big O computational complexity.

### `QueryCombinationIter`

- Moved stuff about `QueryCombinationIter` from `Query` methods to the item itself.

### `WorldQuery` trait

- Introduction has been tidied up
- Macro usage moved to macro item, with a link in the original place.

### `WorldQuery` macro

- Removed foobar terminology in favor of more recognizable identifiers (`ComponentA`, `my_system`). Examples with concrete identifiers are used where appropriate.
- Removed “Ignored fields” section on `WorldQuery` macro.
- Examples split by use case.

## Additional notes

- ~~I had to add `bevy_ecs` as a dev-dependency of `bevy_ecs_macros` to make doctests compile. Discussion on this is appreciated. I'm putting `S-Controversial` for now until a collective decision can be made.~~ [Solved][2].
- `Query` and `WorldQuery` macro idioms should be in the book, but in the meantime they can stay here.
- ~~I don't know what the computational complexity for `many_iter` and `many_for_each_mut` is, so I added a placeholder for now.~~ [Solved][1].
- Consider removing the `#[world_query(ignore)]` feature, as it is no longer needed.
- Please review by checking the rendered docs (`cargo doc -p bevy_ecs --no-deps --open`).
- Suggestions on improving examples by making them simpler or more concrete (where applicable) are appreciated.

### Todo list:

- [x] Resolve the dev-dependency dilemma
- [x] Write the computational complexity for `many_iter` and `many_for_each_mut`

---

- Fixes #5506

[1]: https://github.com/bevyengine/bevy/pull/4989#issuecomment-1152990854
[2]: https://github.com/bevyengine/bevy/pull/4989#issuecomment-1153124121

alice-i-cecile commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,265 @@ use crate::{
+/// |:---:|:---:|
+/// |[`iter`]\([`_mut`][iter_mut])|O(n)|
+/// |[`for_each`]\([`_mut`][for_each_mut]),<br>[`par_for_each`]\([`_mut`][par_for_each_mut])|O(n)|
+/// |[`iter_many`],<br>[`many_for_each_mut`]|O(???)|
```
This needs to be answered before merging.

alice-i-cecile commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,265 @@ use crate::{
 use std::{any::TypeId, borrow::Borrow, fmt::Debug};
 
-/// Provides scoped access to components in a [`World`].
+/// [System parameter] that provides selective access to data stored in a [`World`].
```
```suggestion
/// [System parameter] that provides selective access to the [`Component`] data stored in a [`World`].
```

alice-i-cecile replied:
It doesn't provide access to resources, and this is important to highlight up front.

devil-ira commented:
`many_iter` is essentially a bunch of `Query::get`'s with slightly reduced overhead.
So the complexity is `O(n)` where `n` is the length of the provided list of entities.

alice-i-cecile commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -1075,23 +1177,17 @@ impl<'w, 's, Q: WorldQuery, F: WorldQuery> Query<'w, 's, Q, F> {
     ///
-    /// Panics if the number of query results is not exactly one. Use
-    /// [`get_single`](Self::get_single) to return a `Result` instead of panicking.
+    /// - [`get_single`](Self::get_single) for the non-panicking version.
```
These should probably link to-and-from their mutable twins?

Nilirad replied:
Yes, that was my intention for all the methods. Good catch 👍🏻

alice-i-cecile reviewed (commented):
I really like the consistency this adds, and the language changes are good. There's a few things to clean up. Not sure if / how we can resolve the dev-dependency problem; I'd probably just revert that and merge it without links for now :(

Nilirad commented:
I moved back the macro docs to the `WorldQuery` trait and restored changes to `bevy_ecs_macros` from the `main` branch.

I also removed the `S-Controversial` label.

MinerSebas commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,265 @@ use crate::{
+/// However, iteration cost of queries is reduced by the fact that they have an internal archetype cache to avoid re-computing archetype matches on each query access.
 ///
-/// ## Getting the query result for a particular entity
+/// Two systems cannot be executed in parallel if both access a certain component where at least one of the accesses is mutable.
```
```suggestion
/// Two systems cannot be executed in parallel if both access the same component type where at least one of the accesses is mutable.
```

Nilirad commented on crates/bevy_ecs/src/query/iter.rs:
````diff
@@ -185,6 +185,41 @@ where
+///
+/// This type is returned by calling [`Query::iter_combinations`] and related methods.
+///
+/// It can be iterated by calling [`fetch_next`](Self::fetch_next) in a `while let` loop.
+///
+/// [compiler bug]: https://github.com/rust-lang/rust/issues/62529
+/// [`Query`]: crate::system::Query
+/// [performance section]: crate::system::Query#performance
+/// [`Query::iter_combinations`]: crate::system::Query::iter_combinations
+///
+/// # Example
+///
+/// This example shows how `fetch_next` should be called with a `while let` loop to traverse the iterator.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # #[derive(Component)]
+/// # struct ComponentA;
+/// #
+/// fn some_system(mut query: Query<&mut ComponentA>) {
+///     let mut combinations = query.iter_combinations();
+///     while let Some([a1, a2]) = combinations.fetch_next() {
+///         // ...
+///     }
+/// }
+/// ```
````
It cannot work with the `for` loop because it does not implement the `Iterator` trait.

The older docs in the `iter_combinations` query method specified that this type did not implement `Iterator`, but I don't think it's a good idea to describe an item in terms of what it's not.

At most I can change the first line not use the word “iterator”, as it could be misleading.

devil-ira replied:
[This comment](https://github.com/bevyengine/bevy/pull/4989/files#diff-96cc24e6d6acfeff8249b41cbf81bc003814b93a165eb49bb8edc6b14c1dc668R337-R344) explains why `Iterator` is only implemented for `QueryCombinationIter` with read-only access.

However, the documentation above does not explain this, and mistakenly uses `iter_combinations` instead of `iter_combinations_mut` in the example.

@Nilirad, do you mind clarifying these docs?

Nilirad replied:
Ooh, I completely missed that 🤣 Thanks, i'll make some changes as soon as I can.

Nilirad commented:
Squashed because rebasing in multiple steps is a headache.

I preserved co-authors in the commit message.

james7132 commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -12,94 +12,145 @@ pub use bevy_ecs_macros::WorldQuery;
 use std::{cell::UnsafeCell, marker::PhantomData};
 
-/// Types that can be queried from a [`World`].
+/// Types that can be used to query a [`World`].
```
This introduction is sort of dry. If someone was to happen across this before `Query`, they'd be lost. Providing context as to why this trait exists, or at least link to `Query`  would be a good step forward.  API docs shouldn't have a specific "read-order", and should aggressively link back to core topics so that they're directed there, and dig back deeper if and only if they're required to. `bevy_ecs` is hopelessly pub right now, and `WorldQuery`, while important, is largely an implementation detail with only one outward facing effect: what is and what isn't Queryable.

Nilirad replied:
Thanks for the feedback. I used to have some sort of backlink before, but it was like a duplication of what there is in `Query`.

I restored the backlink by slightly changing the first-line short description.

james7132 commented on crates/bevy_ecs/src/system/query.rs:
````diff
@@ -10,235 +10,265 @@ use crate::{
-/// ```
+/// # Performance
+///
+/// Creating a `Query` is a low-cost operation.
````
```suggestion
/// Creating a `Query` is a `O(1)` operation.
```

Alternatively, low-cost *constant* operation.

Nilirad replied:
I went for the wordy alternative.

james7132 commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,265 @@ use crate::{
+///
+/// Creating a `Query` is a low-cost operation.
+/// Iterating it, on the other hand, fetches data from the world and generates items, which can have a significant computational cost.
+/// However, iteration cost of queries is reduced by the fact that they have an internal archetype cache to avoid re-computing archetype matches on each query access.
```
This is an implementation detail that isn't really relevant in user facing code.

Nilirad replied:
I agree that this sentence contains too many unnecessary details, but I would still like to transmit the “cached” feeling.

james7132 commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,265 @@ use crate::{
+/// # Performance
+///
+/// Creating a `Query` is a low-cost operation.
+/// Iterating it, on the other hand, fetches data from the world and generates items, which can have a significant computational cost.
```
I think it deserves an explicit mention of what kind of storage each component uses will affect the query iteration speed. Sparse set components will force slower archetypal iteration.

james7132 commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,265 @@ use crate::{
+/// |[`single`]\([`_mut`][single_mut]),<br>[`get_single`]\([`_mut`][get_single_mut])|O(a)|
+/// |Archetype based filtering ([`With`], [`Without`], [`Or`])|O(a)|
+/// |Change detection filtering ([`Added`], [`Changed`])|O(a + n)|
+///
```
Probably should add a note here that `for_each` may be conditionally faster to run than `iter` on heavily fragmented worlds, but also discourage it's use as it's generally not as flexible as `iter`.

Nilirad replied:
By heavily fragmented world, you mean a world with a high archetype fragmentation?

alice-i-cecile replied:
That's consistent with my understanding of the performance characteristics, yes.

mockersf replied:
I'm not sure that's still the case? From https://github.com/bevyengine/bevy/pull/4060 it's not obvious at least. In Cart benchmark, when everything is blackboxed, there are no difference.
I don't think that should be mentioned

Nilirad replied:
Hmm, interesting point. Let's see what @james7132 has to say about. If we decide to remove the performance remarks of `for_each`, we should also do the same at the method level.

james7132 replied:
Given that we're still seeing fragmented iteration results that are still at issue here, I still would suggest putting it. From the benchmarks seen while testing #4800, there's still definitely a big difference (though that PR closes the gap significantly), particularly for wider queries that match a lot of archetypes.

james7132 commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,272 @@ use crate::{
+/// |Change detection filtering ([`Added`], [`Changed`])|O(a + n)|
+///
+/// `for_each` methods are seen to be generally faster than their `iter` version on worlds with high archetype fragmentation.
+/// Only benchmarks can guarantee what variant is faster in each use case.
+/// So, if performance analysis is not done, it's better to use the more ergonomic `iter` methods.
```
```suggestion
/// As iterators are in general more flexible and better integrated with the rest of the 
/// Rust ecosystem, it's advised to use `iter` methods over `for_each`. 
/// It is strongly advised to only use `for_each` if it tangibly improves performance: be sure profile
/// or benchmark both before and after the change.
```

Nilirad commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -243,10 +243,10 @@ use std::{any::TypeId, borrow::Borrow, fmt::Debug};
+/// As iterators are in general more flexible and better integrated with the rest of the Rust ecosystem,
+/// it is advised to use `iter` methods over `for_each`.
+/// It is strongly advised to only use `for_each` if it tangibly improves performance:
+/// be sure profile or benchmark both before and after the change.
```
```suggestion
/// be sure to profile or benchmark both before and after the change.
```

Nilirad commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -10,235 +10,274 @@ use crate::{
+/// As iterators are in general more flexible and better integrated with the rest of the Rust ecosystem,
+/// it is advised to use `iter` methods over `for_each`.
+/// It is strongly advised to only use `for_each` if it tangibly improves performance:
+/// be sure profile or benchmark both before and after the change.
```
```suggestion
/// be sure to profile or benchmark both before and after the change.
```

Nilirad commented:
Rebased. I also pushed new commits. The first of the new commits is [5b064a2](https://github.com/bevyengine/bevy/pull/4989/commits/5b064a26d4750a5222301daca046aa7477783f0c).

rparrett commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -9,246 +9,270 @@ use crate::{
+///   The type of data contained in the query item.
+///   Only entities that match the requested data will generate an item.
+/// - **`F` (query filter).**
+///   A set of conditions that determines wheter query items should be kept or discarded.
```
```suggestion
///   A set of conditions that determines whether query items should be kept or discarded.
```

rparrett commented on crates/bevy_ecs/src/system/query.rs:
```diff
@@ -9,246 +9,270 @@ use crate::{
+/// |[`iter_combinations`]\([`_mut`][`iter_combinations_mut`])|Returns an iterator over all combinations of a specified number of query items.|
+/// |[`get`]\([`_mut`][`get_mut`])|Returns the query item for the specified entity.|
+/// |[`many`]\([`_mut`][`many_mut`]),<br>[`get_many`]\([`_mut`][`get_many_mut`])|Returns the query items for the specified entities.|
+/// |[`single`]\([`_mut`][`single_mut`]),<br>[`get_single`]\([`_mut`][`get_single_mut`])|Returns the query item while veryfing that there aren't others.|
```
```suggestion
/// |[`single`]\([`_mut`][`single_mut`]),<br>[`get_single`]\([`_mut`][`get_single_mut`])|Returns the query item while verifying that there aren't others.|
```

rparrett reviewed (commented):
Don't have time to look this over this morning, but I did a quick pass with the spell-checker.

rparrett commented:
Curious about the motivation behind

> Moved stuff about QueryCombinationIter from Query methods to the item itself.

The general pattern in rust docs seems to be for the `Iterator`ish type to point to docs in the methods that create them. `QueryIter` and `QueryManyIter` seem to follow that pattern as well.

Nilirad commented:
> Curious about the motivation behind
> 
> > Moved stuff about QueryCombinationIter from Query methods to the item itself.
> 
> The general pattern in rust docs seems to be for the `Iterator`ish type to point to docs in the methods that create them. `QueryIter` and `QueryManyIter` seem to follow that pattern as well.

I did that because having put that on methods caused [duplication of information](https://dev-docs.bevyengine.org/bevy/ecs/query/struct.QueryState.html#method.iter_combinations), showing that the description is mislocated. This poses docs at a serious risk of remaining outdated in case of changes. This is probably not a piece of information that is going to change, but it's always better to deduplicate documentation and to describe at the nearest point to the implementation.

If you think it could reduce it's visibility, it's much better to link to the iterator instead of reverting the change.

Nilirad commented:
I had to rebase after #5205 got merged, but I don't have enough expertise about `WorldQuery` to fully understand what happened here. So the docs of the `WorldQuery` needs some expert eyes, especially because there is no more `FetchState` and now docs fail to build without warnings. I tried my best but I'm not able to go further alone.

More info on the [commit notes].

[commit notes]: https://github.com/bevyengine/bevy/pull/4989/commits/90a3b16ee6563812131122a53d1720885301584d

alice-i-cecile commented:
Ping @BoxyUwU for docs review following your `WorldQuery` changes.

BoxyUwU commented:
Generally what happend in that PR is that all of the traits were merged into `WorldQuery`. So anything talking about `Fetch::foo` or `FetchState::foo` now should talk about `WorldQuery::foo`

BoxyUwU commented on crates/bevy_ecs/src/query/fetch.rs:
````diff
@@ -12,94 +12,143 @@ pub use bevy_ecs_macros::WorldQuery;
 /// ```
 ///
-/// ## Mutable queries
+/// ## Macro expansion
+///
+/// Expanding the macro will declare three or five additional structs, depending on whether or not the struct is marked as mutable.
+/// For a struct named `X`, the additional structs will be:
+///
+/// |Struct name|`mutable` only|Description|
+/// |:---:|:---:|---|
+/// |`XItem`|---|The type of the query item|
+/// |`XFetch`|---|Implements [`Fetch`] for `XItem`|
+/// |`XState`|---|Implements [`FetchState`]|
+/// |`XReadOnlyItem`|✓|The type of the read-only query item|
+/// |`XReadOnlyFetch`|✓|Implements [`Fetch`] for `XReadOnlyItem`|
+///
````
```suggestion
/// ## Macro expansion
///
/// Expanding the macro will declare three or six additional structs, depending on whether or not the struct is marked as mutable.
/// For a struct named `X`, the additional structs will be:
///
/// |Struct name|`mutable` only|Description|
/// |:---:|:---:|---|
/// |`XState`|---|Used as the [`State`] type for `X` and `XReadOnly`|
/// |`XItem`|---|The type of the query item for `X`|
/// |`XFetch`|---|Used as the [`Fetch`] type for `X`|
/// |`XReadOnlyItem`|✓|The type of the query item for `XReadOnly`|
/// |`XReadOnlyFetch`|✓|Used as the [`Fetch`] type for `XReadOnly`|
/// |`XReadOnly`|✓|ReadOnly variant of `X`|
///
```
Although i am unsure if the `Fetch` and `State` types need to be explained so prominently along with the rest of the information (i.e. `XReadOnly` and `XReadOnlyItem`) since for the most users don't really have to care about these.

BoxyUwU commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -12,94 +12,143 @@ pub use bevy_ecs_macros::WorldQuery;
-/// - Named structs can be clearer and easier to use than complex query tuples. Access via struct
-///   fields is more convenient than destructuring tuples or accessing them via `q.0, q.1, ...`
-///   pattern and saves a lot of maintenance burden when adding or removing components.
-/// - Nested queries enable the composition pattern and makes query types easier to re-use.
-/// - You can bypass the limit of 15 components that exists for query tuples.
+/// component access of `ROQueryFetch<Self>` must be a subset of `QueryFetch<Self>`
+/// and `ROQueryFetch<Self>` must match exactly the same archetypes/tables as `QueryFetch<Self>`
```
This safety comment is already at the bottom of this doc comment so can be removed

BoxyUwU commented on crates/bevy_ecs/src/query/fetch.rs:
````diff
@@ -12,94 +12,143 @@ pub use bevy_ecs_macros::WorldQuery;
+/// ```
+///
+/// To grant mutable access to components, the struct must be marked with the `#[world_query(mutable)]` attribute.
+/// This will also create two more structs that will be used for accessing the query immutably (see table above).
````
```suggestion
/// This will also create three more structs that will be used for accessing the query immutably (see table above).
```

Nilirad commented:
Thanks!!

Nilirad commented:
I split this into multiple PRs, since it's difficult to find people to review it. Hopefully they are more digestible.

- #5739 
- #5740 
- #5741 
- #5742
